File: algorithms/636. Exclusive Time of Functions/solution.cc
class Solution {
public:
    vector<int> exclusiveTime(int n, vector<string>& logs) {
        vector<int> result(n);
        stack<int> s;
        char typestr[10];
        int fnid, time, last_time = 0;
        for ( string & log : logs ) {
            sscanf(log.c_str(), "%d:%[^:]:%d", &fnid, typestr, &time);
            if ( 's' == typestr[0] ) {
                if ( s.size() > 0 )
                    result[s.top()] += time - last_time;
                s.push(fnid);
            } else {
                result[s.top()] += ++time - last_time;
                s.pop();
            }
            last_time = time;
        }
        return result;
    }
};

// 120 / 120 test cases passed.
// Status: Accepted
// Runtime: 45 ms

File: algorithms/636. Exclusive Time of Functions/solution.py
class Solution(object):
    def exclusiveTime(self, n, logs):
        """
        :type n: int
        :type logs: List[str]
        :rtype: List[int]
        """
        result, stack, last_time = [0] * n, [], 0
        for log in logs:
            fn, typestr, time = log.split(':')
            fn, time = int(fn), int(time)
            if typestr == "start":
                if stack:
                    result[stack[-1]] += time - last_time
                stack.append(fn)
                last_time = time
            else:
                result[stack.pop()] += time - last_time + 1
                last_time = time + 1
        return result

# 120 / 120 test cases passed.
# Status: Accepted
# Runtime: 95 ms
